### ISSUE TITLE
[Feature Request]: Pattern: Enriched Document Embedding

### ISSUE THREAD

**Issue by jon-chuang** [jerryjliu/llama_index#6627]
### Feature Description

To have a more robust embedding, I propose the following strategies:

1. Extract title, file name, subsections, other metadata (summary of sub section and article, keyword extraction)
2. Embed a summary of the previous chunk and next chunk.

Problems solved:
1. Coreference resolution / document context 
    - (e.g. multiple 10Ks from the different company over multiple years have the similar sections with similar contents but do not state the year nor company - the metadata or contextual data). 
    - This is a source of retrieval-augmented hallucination.

Strategies for section-wise extraction:
1. Hierarchical
    - Hierarchically construct a summary using the tree-summarizer pattern. This special summary will attempt to hierarchically construct the table of contents by identifying sections and the location of the source documents among them.
2. Linear
    - Attempt to linearly extract subsection titles

In summary, here are the two specific deliverables:
- [ ] Previous and next chunk summary metadata extractor
  - Require `include_prev_next_rel`
- [ ] Title, subsection, keyword metadata extractor

**Comment by jon-chuang** [comment]
TODO:
1. Investigate other methods for long-document understanding.

**Comment by jon-chuang** [comment]
Academic work on summary-based embeddings: https://arxiv.org/abs/2305.12517
